File: server/src/main/java/com/kernith/easyinvoice/data/model/state/quote/QuoteStateFactory.java
package com.kernith.easyinvoice.data.model.state.quote;

import com.kernith.easyinvoice.data.model.Quote;
import com.kernith.easyinvoice.data.model.QuoteStatus;

/**
 * Factory that maps quote status to the appropriate state handler.
 */
public class QuoteStateFactory {
    private QuoteStateFactory() {}

    public static QuoteState from(Quote quote) {
        if (quote == null) {
            throw new IllegalArgumentException("Quote is required");
        }
        return from(quote.getStatus());
    }

    public static QuoteState from(QuoteStatus status) {
        return switch (status) {
            case DRAFT      -> new DraftQuoteState();
            case SENT       -> new SentQuoteState();
            case ACCEPTED   -> new AcceptedQuoteState();
            case REJECTED   -> new RejectedQuoteState();
            case EXPIRED    -> new ExpiredQuoteState();
            case CONVERTED  -> new ConvertedQuoteState();
            case ARCHIVED   -> new ArchivedQuoteState();
        };
    }
}

File: client/src/core/state/quote-state-factory.ts
import { QuoteStatus } from "./quote-status";
import { QuoteState } from "./quote-state";
import {
  AcceptedQuoteState,
  ArchivedQuoteState,
  ConvertedQuoteState,
  DraftQuoteState,
  ExpiredQuoteState,
  RejectedQuoteState,
  SentQuoteState,
} from "./quote-states";

/** Factory that maps quote status to a state handler. */
export class QuoteStateFactory {
  private constructor() {}

  /**
   * @param status current quote status
   */
  static from(status: QuoteStatus): QuoteState {
    switch (status) {
      case QuoteStatus.DRAFT:
        return new DraftQuoteState();
      case QuoteStatus.SENT:
        return new SentQuoteState();
      case QuoteStatus.ACCEPTED:
        return new AcceptedQuoteState();
      case QuoteStatus.REJECTED:
        return new RejectedQuoteState();
      case QuoteStatus.EXPIRED:
        return new ExpiredQuoteState();
      case QuoteStatus.CONVERTED:
        return new ConvertedQuoteState();
      case QuoteStatus.ARCHIVED:
        return new ArchivedQuoteState();
      default:
        return new DraftQuoteState();
    }
  }
}
